flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java

```diff
@@ -0,0 +1,296 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.flink.runtime.scheduler.strategy.SchedulingResultPartition.ResultPartitionState.DONE;
+import static org.apache.flink.runtime.scheduler.strategy.SchedulingResultPartition.ResultPartitionState.PRODUCING;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@link SchedulingStrategy} instance for batch job which schedule vertices when input data are ready.
+ */
+public class LazyFromSourcesSchedulingStrategy implements SchedulingStrategy {
+
+	private final SchedulerOperations schedulerOperations;
+
+	private final SchedulingTopology schedulingTopology;
+
+	private final Map<ExecutionVertexID, DeploymentOption> deploymentOptions;
+
+	private final Map<IntermediateDataSetID, SchedulingIntermediateDataSet> intermediateDataSets;
+
+	public LazyFromSourcesSchedulingStrategy(
+		SchedulerOperations schedulerOperations,
+		SchedulingTopology schedulingTopology) {
+		this.schedulerOperations = checkNotNull(schedulerOperations);
+		this.schedulingTopology = checkNotNull(schedulingTopology);
+		this.intermediateDataSets = new HashMap<>();
+		this.deploymentOptions = new HashMap<>();
+	}
+
+	@Override
+	public void startScheduling() {
+		List<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions = new ArrayList<>();
+		DeploymentOption updateOption = new DeploymentOption(true);
+		DeploymentOption nonUpdateOption = new DeploymentOption(false);
+
+		for (SchedulingExecutionVertex schedulingVertex : schedulingTopology.getVertices()) {
+			DeploymentOption option = nonUpdateOption;
+			for (SchedulingResultPartition srp : schedulingVertex.getProducedResultPartitions()) {
+				SchedulingIntermediateDataSet sid = intermediateDataSets.computeIfAbsent(srp.getResultId(),
+					(key) -> new SchedulingIntermediateDataSet());
+				sid.addSchedulingResultPartition(srp);
+				if (srp.getPartitionType().isPipelined()) {
+					option = updateOption;
+				}
+			}
+			deploymentOptions.put(schedulingVertex.getId(), option);
+
+			if (schedulingVertex.getConsumedResultPartitions().isEmpty()) {
+				// schedule vertices without consumed result partition
+				executionVertexDeploymentOptions.add(
+					new ExecutionVertexDeploymentOption(schedulingVertex.getId(), option));
+			}
+		}
+
+		schedulerOperations.allocateSlotsAndDeploy(executionVertexDeploymentOptions);
+	}
+
+	@Override
+	public void restartTasks(Set<ExecutionVertexID> verticesToRestart) {
+		// increase counter of the dataset first
+		for (ExecutionVertexID executionVertexId : verticesToRestart) {
+			final SchedulingExecutionVertex schedulingVertex = schedulingTopology.getVertex(executionVertexId)
+				.orElseThrow(() -> new IllegalStateException("can not find scheduling vertex for " + executionVertexId));
+
+			for (SchedulingResultPartition srp : schedulingVertex.getProducedResultPartitions()) {
+				if (srp.getPartitionType().isBlocking() && DONE.equals(srp.getState())) {
```

Comment:
If I understand this code correctly, the condition `DONE.equals(srp.getState())` is needed to avoid overcounting in `SchedulingIntermediateDataSet`. Maybe it makes sense in `SchedulingIntermediateDataSet` to keep track of the `IntermediateResultPartitionIDs` that are still being produced to avoid overcounting.

Reply:
Yes, the condition of `DONE.equals(srp.getState())` is to avoid overcounting. One reason of my original thought of not keeping track of `IntermediateResultPartitionIDs` is to save the additional memory of the variables, but now I realize that there may be inconsistency between the counter and the partition state, e.g., restart vertex when partition state is DONE but having not increase the counter, the counter will be over decreased.

I think we should keep track of the `IntermediateResultPartitionIDs` of `DONE` state, which is added in `onExecutionStateChange`. When `restartTasks`, we could use `donePartitionIds.remove(srp.getId())` instead of `DONE.equals(srp.getState())`.

What do you think?